homeassistant/components/go2rtc/__init__.py

```diff
@@ -54,38 +65,75 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    provider = WebRTCProvider(hass, entry.data[CONF_HOST])
     entry.async_on_unload(async_register_webrtc_provider(hass, provider))
     return True
 
 
 class WebRTCProvider(CameraWebRTCProvider):
     """WebRTC provider."""
 
-    def __init__(self, client: Go2RtcClient) -> None:
+    def __init__(self, hass: HomeAssistant,url:str,) -> None:
         """Initialize the WebRTC provider."""
-        self._client = client
+        self._hass = hass
+        self._url = url
+        self._session = async_get_clientsession(hass)
+        self._rest_client = Go2RtcRestClient(self._session, url)
+        self._sessions: dict[str, Go2RtcWsClient] = {}
 
     async def async_is_supported(self, stream_source: str) -> bool:
         """Return if this provider is supports the Camera as source."""
         return stream_source.partition(":")[0] in _SUPPORTED_STREAMS
 
     async def async_handle_web_rtc_offer(
-        self, camera: Camera, offer_sdp: str
-    ) -> str | None:
-        """Handle the WebRTC offer and return an answer."""
-        streams = await self._client.streams.list()
+        self,
+        camera: Camera,
+        offer_sdp: str,
+        session_id: str,
+        send_message: WebRTCSendMessage,
+    ) -> bool:
+        """Handle the WebRTC offer and return the answer via the provided callback.
+
+        Return value determines if the offer was handled successfully.
+        """
+        self._sessions[session_id] = ws_client = Go2RtcWsClient(
+            self._session, self._url, source=camera.entity_id
+        )
+
+        streams = await self._rest_client.streams.list()
         if camera.entity_id not in streams:
             if not (stream_source := await camera.stream_source()):
-                return None
-            await self._client.streams.add(camera.entity_id, stream_source)
+                return False
+            await self._rest_client.streams.add(camera.entity_id, stream_source)
 
-        answer = await self._client.webrtc.forward_whep_sdp_offer(
-            camera.entity_id, WebRTCSdpOffer(offer_sdp)
-        )
-        return answer.sdp
+        @callback
+        def on_messages(message: ReceiveMessages) -> None:
+            """Handle messages."""
+            value: WebRTCMessages
+            if isinstance(message, WebRTCCandidate):
+                value = HAWebRTCCandidate(message.candidate)
+            elif isinstance(message, WebRTCAnswer):
+                value = HAWebRTCAnswer(message.answer)
+
+            send_message(value)
+
+        ws_client.subscribe(on_messages)
+        await ws_client.send(WebRTCOffer(offer_sdp))
+
+        return True
+
+    async def async_on_webrtc_candidate(self, session_id: str, candidate: str) -> None:
+        """Handle the WebRTC candidate."""
+
+        if ws_client := self._sessions.get(session_id):
+            await ws_client.send(WebRTCCandidate(candidate))
+        else:
+            raise ValueError("Unknown session")
```

Comment:
Since closing is happening inside a task, I wouldn't be surprised if this is going to happen quite often (people accidentally opening a camera more info, closing it right away). I wonder if we should just debug log this but not raise ?